Source/Interceptor.cs

```diff
@@ -75,56 +74,50 @@ internal void VerifyAll()
 
 		private void VerifyOrThrow(Func<IProxyCall, bool> match)
 		{
-			lock (calls)
+			var failures = new List<IProxyCall>();
+
+			// The following verification logic will remember each processed setup so that duplicate setups
+			// (that is, setups overridden by later setups with an equivalent expression) can be detected.
+			// To speed up duplicate detection, they are partitioned according to the method they target.
+			var verifiedSetupsPerMethod = new Dictionary<MethodInfo, List<Expression>>();
+
+			foreach (var setup in this.InterceptionContext.OrderedCalls)
 			{
-				var failures = calls.Values.Where(match);
-				if (failures.Any())
+				if (setup.IsConditional)
 				{
-					throw new MockVerificationException(failures.ToArray());
+					continue;
 				}
-			}
-		}
 
-		public void AddCall(IProxyCall call, SetupKind kind)
-		{
-			var expr = call.SetupExpression.PartialMatcherAwareEval();
-			var keyText = call.Method.DeclaringType.FullName + "::" + expr.ToStringFixed(true);
-			if (kind == SetupKind.PropertySet)
-			{
-				keyText = "set::" + keyText;
-			}
+				List<Expression> verifiedSetupsForMethod;
+				if (!verifiedSetupsPerMethod.TryGetValue(setup.Method, out verifiedSetupsForMethod))
+				{
+					verifiedSetupsForMethod = new List<Expression>();
+					verifiedSetupsPerMethod.Add(setup.Method, verifiedSetupsForMethod);
+				}
```

Comment:
There's another occurrence of this code pattern somewhere else in Moq's codebase, perhaps consider creating some kind of `MultiMap` / `MultiDictionary` ADT so this can be abstracted away.